docs/petr4spec/Petr4-spec.mdk

```diff
@@ -142,509 +149,3025 @@ p4grammar {
+- Direction is optional in parameter declaration. However, for simpilicity, we just use $\less$ to denote a missing direction. 
+
+~ Begin P4Syntax
+**Data types used in P4 programs:**
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\baseTyp$ | $::=$ | $\boolTyp$ | $\qquad \text{booleans}$ |
+|   | $\vert$ | $\errTyp$ | $\qquad \text{errors}$ |
+|   | $\vert$ | $\integerTyp$ | $\qquad \text{infinite-precision integers}$ |
+|   | $\vert$ | $\intWidthTyp \exp$ | $\qquad \text{fixed-width signed integers}$ |
+|   | $\vert$ | $\bitWidthTyp \exp$ | $\qquad \text{fixed-width unsigned integers}$ |
+|   | $\vert$ | $\varBitTyp \exp$ | $\qquad \text{variable-width integers with a maximum width}$ |
+|   | $\vert$ | $\typVar$ | $\qquad \text{type variables}$ |
+|   | $\vert$ | $\spcTyp \baseTyp \baseTyps$ | $\qquad \text{specialized type}$ |
+|   | $\vert$ | $\stackTyp \baseTyp \exp$ | $\qquad \text{stacks}$ |
+|   | $\vert$ | $\tupleTyps \baseTyps$ | $\qquad \text{tuples}$ |
+|   | $\vert$ | $\stringTyp$ | $\qquad \text{strings}$ |
+|   | $\vert$ | $\voidTyp$ | $\qquad \text{void}$ |
+|   | $\vert$ | $\dontcareTyp$ | $\qquad \text{don't care}$ |
+{ .booktable }
+
+**Directions:**
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\dir$ | $::=$ | $\inDir$ | $\qquad \text{copy-in}$ |
+|   | $\vert$ | $\out$ | $\qquad \text{copy-out}$ |
+|   | $\vert$ | $\inout$ | $\qquad \text{copy-in-out}$ |
+|   | $\vert$ | $\less$ | $\qquad \text{direction-less}$ |
+{ .booktable }
+
+**Parameters:**
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\prmDef \typKind$ | $::=$ | $\dir \, \typKind \, \var$ | $\qquad \text{parameters}$ |
+|   | $\vert$ | $\dir \, \typKind \, \var = \exp$ | $\qquad \text{parameters with value}$ |
+{ .booktable }
+
+
+**Unary operations:**
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\unaryOp$ | $::=$ | $!$ | $\qquad \text{logical negation}$ |
+|   | $\vert$ | $\bitComplement$ | $\qquad \text{bitwise complement}$ |
+|   | $\vert$ | $-$ | $\qquad \text{unary minus}$ |
+{ .booktable }
+
+**Binary operations:**
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\binOp$ | $::=$ | $\&\&$ | $\qquad \text{logical and}$ |
+|   | $\vert$ | $\vert\vert$ | $\qquad \text{logical or}$ |
+|   | $\vert$ | $+$ | $\qquad \text{numerical addition}$ |
+|   | $\vert$ | $-$ | $\qquad \text{numerical subtraction}$ |
+|   | $\vert$ | $*$ | $\qquad \text{numerical multiplication}$ |
+|   | $\vert$ | $\div$ | $\qquad \text{numerical division}$ |
+|   | $\vert$ | $\mod$ | $\qquad \text{numerical modulo}$ |
+|   | $\vert$ | $==$ | $\qquad \text{equality check}$ |
+|   | $\vert$ | $!=$ | $\qquad \text{inequality check}$ |
+|   | $\vert$ | $\plusSat$ | $\qquad \text{saturating addition}$ |
```

Comment:
These operators might look better in \mathsf or \textt.

Reply:
It definitely looks nicer however it doesn't work with $\sim$, $\leq$, $\geq$, $\gg$, $\ll$, and $\hat{}$. Figure a way around it when it comes to review and clean up.

Reply:
Ooh, that's annoying. I think `|+|` and other saturating operations are probably best rendered either as code font `|+|` or as some kind of TeX glyph like \boxplus.

Sorry if this is sending you emails btw I know it's 11 o clock.

Reply:
I'll try \boxplus. and no worries, I appreciate your feedback any time of the day ;)

Reply:
figured it out.